ADAL/src/workplacejoin/mac/ADWorkPlaceJoinUtil.m

```diff
@@ -22,14 +22,292 @@
+    }
+    
+    // If there's no certificate in the keychain, return nil. adError won't be set if the
+    // cert can't be found since this isn't considered an error condition.
+    if (!certificate)
+        return nil;
+    
+    certificateSubject = (__bridge_transfer NSString*)(SecCertificateCopySubjectSummary(certificate));
+    certificateData = (__bridge_transfer NSData*)(SecCertificateCopyData(certificate));
+    
+    
+    // Get the private key
+    AD_LOG_VERBOSE(@"Retrieving WPJ private key reference", correlationId, nil);
+    
+    privateKey = [self copyPrivateKeyRefForIdentifier:privateKeyIdentifier correlationId:correlationId error:&adError];
+    if (adError)
+    {
+        if (error)
+            *error = adError;
+        AD_LOG_ERROR_F(@"Failed to retrieve WPJ private key reference - ", adError.code, nil, @"%@ correlation id.", correlationId);
+        goto _error;
+    }
+    
+    
+    // Get user principal name
+    AD_LOG_VERBOSE(@"Retrieving WPJ user principal name", correlationId, nil);
+    
+    userPrincipalName = [self stringDataFromIdentifier:upnIdentifier correlationId:correlationId error:&adError];
+    if (adError)
+    {
+        if (error)
+            *error = adError;
+        AD_LOG_ERROR_F(@"Failed to retrieve WPJ user principal name from the keychain - ", adError.code, nil, @"%@ correlation id.", correlationId);
+        goto _error;
+    }
+    
+    if (!identity || !userPrincipalName || !certificateIssuer || !certificateSubject || !certificateData || !privateKey)
+    {
+        // The code above will catch missing security items, but not missing item attributes. These are caught here.
+        ADAuthenticationError* adError = [ADAuthenticationError unexpectedInternalError:@"Missing some piece of WPJ data" correlationId:correlationId];
+        if (error)
+            *error = adError;
+        goto _error;
+    }
+    
+    // We found all the required WPJ information.
+    info = [[ADRegistrationInformation alloc] initWithSecurityIdentity:identity
+                                                     userPrincipalName:userPrincipalName
+                                                     certificateIssuer:certificateIssuer
+                                                           certificate:certificate
+                                                    certificateSubject:certificateSubject
+                                                       certificateData:certificateData
+                                                            privateKey:privateKey];
+    SAFE_ARC_AUTORELEASE(info);
+    
+    // Fall through to clean up resources.
+    
+_error:
+    
+    SAFE_ARC_RELEASE(certificateSubject);
+    SAFE_ARC_RELEASE(certificateData);
+    
+    if (identity)
+        CFRelease(identity);
+    if (certificate)
+        CFRelease(certificate);
+    if (privateKey)
+        CFRelease(privateKey);
+    
+    return info;
+}
+
+
++ (void) copyCertificate:(NSUUID*)correlationId
```

Comment:
The certificate: shouldn't be repeated twice in this method signature. The certificate double pointer should be passed in here, not correlation ID. Also, please take in a id<ADRequestContext> instead of correlationId, so that way we can get the telemetry information as well.